backend/server/routers/courses.py

```diff
@@ -136,23 +180,47 @@ def search(string):
           “COMP1531”: “SoftEng Fundamentals,
             ……. }
     """
-    # TODO: is regex search really something we want?
-    # malicious regex can cause DOS depending on regex implementation
-    # Would fuzzy search be better?
-    pat = re.compile(r"{}".format(string), re.I)
-    code_query = list(coursesCOL.find({"code": {"$regex": pat}}))
-    title_query = list(coursesCOL.find({"title": {"$regex": pat}}))
-
-    if not code_query and not title_query:
+    global ALL_COURSES
+    from server.routers.programs import getStructure
+
+    if ALL_COURSES is None:
+        ALL_COURSES = fetch_all_courses()
+
+    # TODO: can you have a minor without a major selected?
+    #       will wreak havoc on the argument order with *specialisations
+    #       currently this is enforced during setup but will need to
+    #       make sure that this is true for all degrees not just comp sci
+    specialisations = list(userData.specialisations.keys())
+    structure = getStructure(userData.program, *specialisations)['structure']
+
+    top_results = sorted(ALL_COURSES.items(), reverse=True,
+                         key=lambda course: fuzzy_match(course, search_string)
+                         )[:100]
+    weighted_results = sorted(top_results, reverse=True,
+                              key=lambda course: weight_course(course,
+                                                               search_string,
+                                                               structure, 
+                                                               *specialisations)
+                              )[:30]
+
+    return {course[0]: course[1] for course in weighted_results}
```

Comment:
This is mean, but you should use tuple decomposition here.

Reply:
Wasn't sure how to do that in a list comprehension

Reply:
```python
return {code: name for code, name in weighted_results}
```

Reply:
oh yeah im just dumb